File: AlgorithmLcr/Try1-Completely unknown single port fitting/rlc_id/report.py
"""ASCII identification report (DESIGN.md §5.5 D6.4, §8.1 metrics)."""

from __future__ import annotations

import numpy as np

from .circuits import to_string
from .selector import EquivalenceClass


def _fmt(x: float, width: int = 10, prec: int = 4) -> str:
    if not np.isfinite(x):
        return f"{'inf' if x > 0 else '-inf':>{width}}"
    return f"{x:>{width}.{prec}g}"


def format_report(title: str, classes: list[EquivalenceClass],
                  truth: str | None = None, top_k: int = 5,
                  extra_lines: list[str] | None = None) -> str:
    """Ranked candidate table: rank / topology+values / n / wRMSE / maxRelErr
    / AICc / equivalence annotation."""
    lines: list[str] = []
    lines.append("=" * 96)
    lines.append(f"DUT: {title}")
    if truth:
        lines.append(f"truth: {truth}")
    lines.append("-" * 96)
    header = (f"{'rk':>3} {'topology':<46} {'n':>2} {'wRMSE':>10} "
              f"{'maxRelErr':>10} {'AICc':>10}  note")
    lines.append(header)
    lines.append("-" * 96)
    if not classes:
        lines.append("(no valid candidates)")
    aicc0 = classes[0].aicc if classes else 0.0
    for rank, eq in enumerate(classes[:top_k], start=1):
        rep = eq.representative
        topo = to_string(rep.tree, rep.theta)
        if len(topo) > 46:
            topo = topo[:43] + "..."
        daicc = rep.aicc_val - aicc0
        notes: list[str] = []
        if len(eq.members) > 1:
            notes.append(f"equiv x{len(eq.members)}")
        if 0 < daicc < 2.0:
            notes.append("dAICc<2")
        if rep.engine == "B":
            notes.append("engB")
        if rep.note:
            notes.append(rep.note)
        lines.append(
            f"{rank:>3} {topo:<46} {rep.n_params:>2} "
            f"{_fmt(rep.wrmse)} {_fmt(rep.max_rel_err)} {_fmt(rep.aicc_val, 10, 2)}"
            f"  {'; '.join(notes)}")
        alt_count = 0
        for member in eq.members:
            if member is rep:
                continue
            if alt_count >= 2:
                break
            alt_count += 1
            mtopo = to_string(member.tree, member.theta)
            if len(mtopo) > 40:
                mtopo = mtopo[:37] + "..."
            lines.append(f"    └─ equiv: {mtopo:<40} (eng{member.engine}, "
                         f"AICc {member.aicc_val:.2f})")
    if extra_lines:
        lines.append("-" * 96)
        lines.extend(extra_lines)
    lines.append("=" * 96)
    return "\n".join(lines)


def format_skipped(skipped_notes: list[str]) -> str:
    """Format engine-B D8 skip annotations."""
    if not skipped_notes:
        return ""
    return "engine-B skipped candidates (D8): " + " | ".join(skipped_notes)

File: AlgorithmLcr/cppversion/src/report.cpp
#include "report.hpp"

#include <cmath>
#include <cstdio>

namespace rlc {

namespace {
std::string fmt(double x, int width = 10, int prec = 4) {
    if (!std::isfinite(x)) {
        std::string s = x > 0 ? "inf" : "-inf";
        while ((int)s.size() < width) s = " " + s;
        return s;
    }
    char buf[64];
    std::snprintf(buf, sizeof(buf), "%*.*g", width, prec, x);
    return buf;
}
}  // namespace

std::string formatReport(const std::string& title,
                         const std::vector<EquivalenceClass>& classes,
                         const std::string* truth, int topK,
                         const std::vector<std::string>* extraLines) {
    std::vector<std::string> lines;
    std::string bar(96, '=');
    std::string dash(96, '-');
    lines.push_back(bar);
    lines.push_back("DUT: " + title);
    if (truth) lines.push_back("truth: " + *truth);
    lines.push_back(dash);
    {
        char buf[128];
        std::snprintf(buf, sizeof(buf), "%3s %-46s %2s %10s %10s %10s  note", "rk",
                      "topology", "n", "wRMSE", "maxRelErr", "AICc");
        lines.push_back(buf);
    }
    lines.push_back(dash);
    if (classes.empty()) lines.push_back("(no valid candidates)");
    double aicc0 = classes.empty() ? 0.0 : classes[0].aicc();
    int rank = 1;
    for (const auto& eq : classes) {
        if (rank > topK) break;
        const Candidate& rep = eq.representative;
        std::string topo = toString(rep.tree, &rep.theta);
        if (topo.size() > 46) topo = topo.substr(0, 43) + "...";
        double daicc = rep.aiccVal - aicc0;
        std::vector<std::string> notes;
        if (eq.members.size() > 1) {
            char buf[32];
            std::snprintf(buf, sizeof(buf), "equiv x%zu", eq.members.size());
            notes.push_back(buf);
        }
        if (daicc > 0 && daicc < 2.0) notes.push_back("dAICc<2");
        if (rep.engine == "B") notes.push_back("engB");
        if (!rep.note.empty()) notes.push_back(rep.note);
        std::string noteJoin;
        for (size_t i = 0; i < notes.size(); ++i) {
            if (i) noteJoin += "; ";
            noteJoin += notes[i];
        }
        char buf[160];
        std::snprintf(buf, sizeof(buf), "%3d %-46s %2d %s %s %s  %s", rank, topo.c_str(),
                      rep.nParams(), fmt(rep.wrmse).c_str(), fmt(rep.maxRelErr).c_str(),
                      fmt(rep.aiccVal, 10, 2).c_str(), noteJoin.c_str());
        lines.push_back(buf);
        int altCount = 0;
        for (const auto& member : eq.members) {
            if (&member == &rep) continue;
            if (altCount >= 2) break;
            ++altCount;
            std::string mtopo = toString(member.tree, &member.theta);
            if (mtopo.size() > 40) mtopo = mtopo.substr(0, 37) + "...";
            char b2[128];
            std::snprintf(b2, sizeof(b2), "    + equiv: %-40s (eng%s, AICc %.2f)",
                          mtopo.c_str(), member.engine.c_str(), member.aiccVal);
            lines.push_back(b2);
        }
        ++rank;
    }
    if (extraLines) {
        lines.push_back(dash);
        for (const auto& l : *extraLines) lines.push_back(l);
    }
    lines.push_back(bar);
    std::string out;
    for (size_t i = 0; i < lines.size(); ++i) {
        if (i) out += "\n";
        out += lines[i];
    }
    return out;
}

}  // namespace rlc
